File: 61-rotate-list/61-rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        
        if ( !(head) || !(head -> next) ) return head;
        
        ListNode* fw = head;
        ListNode* bw = head;
        int len = 0;
        int ahead = k;
        
        while (ahead--) {
            len += 1;
            
            if ( !(fw -> next) ) {
                ahead = k % len;
                fw = head;
                break;
            }
            
            fw = fw -> next;
        }
        
        if (ahead == 0) return head;
        
        while (ahead > 0) {
            fw = fw -> next;
            ahead--;
        }
        
        while (fw -> next) {
            fw = fw -> next;
            bw = bw -> next;
        }
        
        fw -> next = head;
        head = bw -> next;
        bw -> next = nullptr;
        
        return head;
    }
};

File: 61-rotate-list/61-rotate-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var rotateRight = function(head, k) {
    
    if (!head || !(head.next) ) return head;
    
    let fw = head;
    let bw = head;
    
    while (k--) fw = fw.next ? fw.next : head;
    
    while (fw.next) {
        fw = fw.next
        bw = bw.next
    }
    
    fw.next = head;
    head = bw.next;
    bw.next = null
    
    return head;
    
};